python/podio_gen/generator_utils.py

```diff
@@ -29,6 +29,37 @@ def _prefix_name(name, prefix):
   return name
 
 
+def get_julia_type(cpp_type=None, is_array=False, array_type=None, array_size=None):
+  """Parse the given c++ type to a Julia type"""
+  builtin_types_map = {"int": "Int32", "float": "Float32", "double": "Float64",
+                       "bool": "Bool", "long": "Int64", "unsigned int": "UInt32",
+                       "unsigned long": "UInt64", "char": "Char", "short": "Int16",
+                       "long long": "Int64", "unsigned long long": "UInt64"}
+  # is a global type as described in test_MemberParser.py #L121
+  if cpp_type and cpp_type.startswith("::"):
```

Comment:
```suggestion
  if cpp_type.startswith("::"):
```
No need to check here, since we will always have a value.

Reply:
@tmadlener Can't make this change as it is throwing error in build itself for example datamodel.
```
File "/home/user/podio/python/podio_gen/generator_utils.py", line 39, in get_julia_type
    if cpp_type.startswith("::"):
AttributeError: 'NoneType' object has no attribute 'startswith'
CMake Error at cmake/podioMacros.cmake:219 (message):
  Could not generate datamodel 'datamodel'.  Check your definition in
  'datalayout.yaml'
Call Stack (most recent call first):
  tests/CMakeLists.txt:8 (PODIO_GENERATE_DATAMODEL)
```

Reply:
This looks a bit like a logic error somewhere, as just from the calling sites there should not be any way to call this with a `cpp_type` that is `None`. Is that the complete stacktrace you got? It is hard to tell just from this bit where the original call came from.

Reply:
I feel there is an else block too that's checking for array type. So if cpp_type is None, it will not work for all the three ifs and we have to pass the array_type present in else condition.

```
if cpp_type and cpp_type.startswith("::"):
    cpp_type = cpp_type[2:]
  if cpp_type in builtin_types_map:
    return builtin_types_map[cpp_type]

  if not is_array:
    if cpp_type.startswith('std::'):
      cpp_type = cpp_type[5:]
    if cpp_type in ALLOWED_FIXED_WIDTH_TYPES:
      regex_string = re.split("(u|)int(8|16|32|64)_t", cpp_type)
      cpp_type = regex_string[1].upper() + "Int" + regex_string[2]
      return cpp_type

  else:
    array_type = get_julia_type(cpp_type=array_type)
    if '::' in array_type:
      array_type = array_type.split('::')[1]
    if array_type not in builtin_types_map.values():
      array_type = array_type + 'Struct'
    return f"MVector{{{array_size}, {array_type}}}"

  return cpp_type
```

Reply:
This member: `std::array<int, 4> p [mm]` caused the error. I think that's why we have an else condition that handles `array_type` like these and for such members `cpp_type` is None.

Reply:
OK. That is slightly weird as at least from the parsing point of view, `std::array<int, 4> p [mm]` should definitely parse into a `MemberVariable` that has `array_type` set. I have just briefly checked the unit tests in [`test_MemberParser.py`](https://github.com/AIDASoft/podio/blob/master/python/podio_gen/test_MemberParser.py). It looks like we do not yet have an explicit test for this, so maybe this is indeed not working as intended.

I think the quickest test would be to add a unittest using exactly this type and see if fails, i.e. add something similar to, e.g. this:
https://github.com/AIDASoft/podio/blob/858c0ff0b841705d1b18aafd57569fcbd1beda91/python/podio_gen/test_MemberParser.py#L230-L234

to the `test_parse_unit` function using `std::array<int, 4> p [mm]` as input and see if it works. (You can run the python unittests even without a successul build, via `python test_MemberParser.py`)

Reply:
This is how the tests would look like:
```
    parsed = parser.parse('std::array<int, 4> p [mm]', False)
    self.assertEqual(parsed.full_type, 'std::array<int, 4>')
    self.assertEqual(parsed.name, 'p')
    self.assertEqual(parsed.array_type, 'int')
    self.assertTrue(parsed.is_builtin_array)
```
I am unable to run the tests though. Getting this error.
```
from podio_gen.podio_config_reader import MemberParser, DefinitionError
ModuleNotFoundError: No module named 'podio_gen'
```

Reply:
I just ran them quickly o my end, and this is working. So the error has to be in somewhere in the whole chain of calling things, such that the `arra_type` is either overwritten, or you end up calling `get_julia_type` with something where you should not call it.

You might have to run the tests in the `python` subdirectory via `python -m unittest podio_gen/test_MemberParser.py`

Reply:
I'll see what's going wrong.

Reply:
Ah right, that makes sense (sort of at least). I have the feeling that the `is_array` argument is redundant in this case as `bare_type` will only be `None` in case of `std::array` variables, but I am not entirely sure. If you want you can try to see if that is actually the case. However, I would be happy with the following (much less time consuming) as well:

- Make the argument required (i.e. remove the `=None` from `cpp_type` in the argument list)
- Keep the check in the function body as it is and add a comment that summarizes what you have figured out in this thread (something like, cpp_type can be None in case of array members)

Reply:
Okay. And for the other two ?
>array_type = get_julia_type(array_type)
>self.julia_type = get_julia_type(self.bare_type, is_array=self.is_array,